### ISSUE TITLE
[Bug] Motion Control not Working (Extremely Buggy) for Switch Pro Controller in Zelda - Breath of the Wild

### ISSUE THREAD

**Issue by Light-Blue-Yoshi** [Ryujinx/Ryujinx#4883]
### Description of the issue

I have two top tier 3rd party controllers both capable of running in dedicated Nintendo Switch Pro Controller mode and with motion control (gyroscope) function that works in Yuzu for motion control aim assist but doesn't work (or is extremely buggy) in Ryujinx. The game tested on is The Legend of Zelda - Breath of the Wild and also another game with the exact same problem. The controllers tested are the GuliKit KingKong 2 Pro Controller and the GuiliKit KingKong 2 Controller. Both controllers have the function (selected by pressing mode button on controller) to to connect via Bluetooth in either of these 4 modes:

Nintendo Switch Pro Controller Mode
Windows PC XINPUT Mode
Android & iOS Mode
Windows (Older) PC DINPUT Mode

Using Windows 11, I have tried Bluetooth pairing these 2 controllers in "Nintendo Switch Pro Controller Mode" and also tried in "Windows PC XINPUT Mode". Both controllers work with in these modes for most functions except I find that so far the only way I can get motion control to work is with both controllers connected in "Nintendo Switch Pro Controller Mode" and then only reliably with motion control using Yuzu Emulator and not Ryujinx. When I use the controllers in either Switch Pro Controller Mode or Windows PC XINPUT Mode there is not motion control in Ryujinx.

So here is the really weird buggy thing is if I connect the controllers in "Nintendo Switch Pro Controller Mode" and then run Zelda - Breath of the Wild on Yuzu (where motion control aiming assist works fine), close the game but leave Yuzu open and then open the game in Ruyjinx then all of a sudden the motion control aiming works sometimes in Ryujinx and then will sometimes stop working again. If I close Yuzu and then open Ryujinx and then Zelda - Breath of the Wild then the pro controller usually has no more motion control aiming assist.

I'm using the latest Ryujinx and Yuzu Mainline version. Please note that since these top tier 3rd party GuliKit Controllers are made to connect directly to a Nintendo Switch as a Pro Controller and also show up in Ryujinx as "Nintendo Switch Pro Controller" I would assume that the genuine 1st party Nintendo Swtich Pro Controllers will most likley have exactly the same issue of no motion control aim assist in Zelda - Breath of the Wild (or most likley other games) in Ryujinx.

I would be most grateful if you could please fix this as Ryujinx IMHO is a far better emulator to run most games as I find Yuzu has significant more lags and bugs even on my brand $4500 water cooled gaming laptop from 2023 with the latest spec intel i9 24-core CPU, nVidia RTX 4070, 32GB 5600 DDR4 RAM & 2TB Kingston Fury M.2 NVMe SSD (7GB/s) running latest vers Win11 & all recent drivers.

### Reproduction steps

- Open Ruyjinx.
- Go to Options, Settings, Input tab & Click on Player 1 Configure. Docked Mode it also on.
- Make sure sure Nintendo Pro Controller is selected and setup correctly with the optio "Enable Motion Control" selected.
- Run the game The Legend of Zelda - Breath of the Wild.
- Either load a savegame or if starting new, progress to the point where you exit the cave and pick up at least one weapon such as a wooden branch.
- In the game options (press the plus button [+] on controller and use R and L to navigate) and make sure the option for motion control aiming assist is on.
- In the game, you can try either of these two options to test if motion control aiming assist is working or not. Either:
   1) Press the right joystick button to use the zoom camera. You can now move the controller up, down and turn it from side to side when motion control aiming assist is working the whole screen will move and aim. When it's not working it will not move at all and you can only aim with the right joysick.
   2) When equipt with a weapon such as a tree branch, wooden club, sword, ect you can press and hold R shoulder button. You can now move the controller up, down and turn it from side to side when motion control aiming assist is working the whole screen will move and aim. When it's not working it will not move at all and you can only aim with the right joysick.

### Log file

[Ryujinx_1.1.777_2023-05-11_13-10-45.log](https://github.com/Ryujinx/Ryujinx/files/11447900/Ryujinx_1.1.777_2023-05-11_13-10-45.log)

### OS

Windows 11

### Ryujinx version

1.1.777

### Game version

1.6.0

### CPU

Intel i9-13900HX

### GPU

NVIDIA RTX 4070

### RAM

32GB DDR4 (5600 Speed)

**Comment by OldManKain** [comment]
I do not have this game to test but I do have 3 GuliKit King King 2 Pro/Non-Pro controllers that I just tested on MK8D that work and have not had any issues in motion control games on Ryujinx.

GuliKit controllers only have true Gyro in Switch mode, also in switch mode set your controller on a flat surface to calibrate the gyro with “+,-, A and D-pad Left”

Make sure you are on the latest Controller Firmware v4.9 and controller CPU Firmware v1.17

**Comment by Light-Blue-Yoshi** [comment]
OldManKain - Thanks for your input. Please can you expand on what MK8D is? Is that Mario Kart 8 Deluxe? If so, I have the latest version of this game installed and would be most grateful if you could explain how to test motion control in the game?

I had already done the step above of placing on a flat table and doing "+ and - and A and D-pad left" to reset the gyros on both controllers as well as "L and R and A and D-pad left" to reset sticks and trigger buttons. None of these made a difference so I also also tried "Gear button and -" to wipe all settings.

Please note that GuliKit says that Controller Firmware v4.9 is only for changing the operation very slightly for resetting the calibration of sticks and trigger buttons. The combo of  "L and R and A and D-pad left" to reset sticks and trigger buttons. Before update you'd only need to press these buttons and it would instantly vibrate and reset. After the update it changes so you have to hold that button combo for 2sec to get it to vibrate and then reset the sticks and trigger buttons. So it really shouldn't apply to anything here regarding the gyro functions but I had already had the same bug in Ruyjinx before and after applying this firmware update.

The last update you mentioned CPU Firmware v1.17 well accordingto GuliKit it says it only for:
 "Optimize the NFC driver to support some non-standard NFC cards". Since the controller supports reading Amiibo I assume this is what it applys to. In any case this firmware update also doesn't seem to have any applicable relation to the bug I described such as function of the controlls such as the gyro.

**Comment by piplup55** [comment]
can't reproduce, needs further testing

**Comment by OldManKain** [comment]
> OldManKain - Thanks for your input. Please can you expand on what MK8D is? Is that Mario Kart 8 Deluxe? If so, I have the latest version of this game installed and would be most grateful if you could explain how to test motion control in the game?
> 
> I had already done the step above of placing on a flat table and doing "+ and - and A and D-pad left" to reset the gyros on both controllers as well as "L and R and A and D-pad left" to reset sticks and trigger buttons. None of these made a difference so I also also tried "Gear button and -" to wipe all settings.
> 
> Please note that GuliKit says that Controller Firmware v4.9 is only for changing the operation very slightly for resetting the calibration of sticks and trigger buttons. The combo of "L and R and A and D-pad left" to reset sticks and trigger buttons. Before update you'd only need to press these buttons and it would instantly vibrate and reset. After the update it changes so you have to hold that button combo for 2sec to get it to vibrate and then reset the sticks and trigger buttons. So it really shouldn't apply to anything here regarding the gyro functions but I had already had the same bug in Ruyjinx before and after applying this firmware update.
> 
> The last update you mentioned CPU Firmware v1.17 well accordingto GuliKit it says it only for: "Optimize the NFC driver to support some non-standard NFC cards". Since the controller supports reading Amiibo I assume this is what it applys to. In any case this firmware update also doesn't seem to have any applicable relation to the bug I described such as function of the controlls such as the gyro.

Yes Mario Kart 8 Deluxe you can enable motion controls by clicking this or pressing Y
![image](https://github.com/Ryujinx/Ryujinx/assets/47503099/bb5cab86-5f68-48f0-bcd7-6e1dd6c82cf2)


As for the firmware's I had to update them to 1.1.6 and 4.5 before motion controls would work. they come preloaded with really old firmware's depending how long they have been sitting on warehouse shelves.  Never hurts to update.

**Comment by Light-Blue-Yoshi** [comment]
OldManKain - While I remember this may help you... I noticed last night that some very interesting bugs and found a workaround for getting both of my GuliKit KingKong 2 Pro and non pro controllers to work with prober vibration feedback in Mario Kart 8 Deluxe. Here is what I found in Mario Kart 8 Deluxe:

- Both Controllers connected in Switch mode, No vibration feedback in Ryujinx.
- Both Controllers connected in Windows X-Input mode, No Viration feedback in Yuzu.
- Both Controllers connected in Switch mode, No vibration feedback in Yuzu for the Pro controller but really strong vibration feedback for the non pro controller. (I tried adjusting vibration with "Gear Button and D-pad Up" on the pro controller but still Yuzu didn't want to make it work. The non pro controller I tried turning the vibration down with "Gear Button and D-pad Down" and it went from too strong vibration to kinda too weak.
- Both Controllers connected in Windows X-Input mode now vibration works correctly for both controllers. I find I like it just a little bit stronger rumble so I press "Gear Button and D-pad Up" so it's on vibration level 3 for my preferred setting. :)

I only briefly tested Yuzu and Ruyjinx today with Crash Team Racing Nitro-Fueled. It seems again the way to get both controllers to work with vibration feedback is connected in Windows X-Input mode and only on Ryujinx emulator. Seriously I like the layout of Yuzu but I am really annoyed how much time I've wasted with bugs on Yuzu only to find and Ryujinx works better for most things.

**Comment by OldManKain** [comment]
> OldManKain - While I remember this may help you... I noticed last night that some very interesting bugs and found a workaround for getting both of my GuliKit KingKong 2 Pro and non pro controllers to work with prober vibration feedback in Mario Kart 8 Deluxe. Here is what I found in Mario Kart 8 Deluxe:
> 
> * Both Controllers connected in Switch mode, No vibration feedback in Ryujinx.
> * Both Controllers connected in Windows X-Input mode, No Viration feedback in Yuzu.
> * Both Controllers connected in Switch mode, No vibration feedback in Yuzu for the Pro controller but really strong vibration feedback for the non pro controller. (I tried adjusting vibration with "Gear Button and D-pad Up" on the pro controller but still Yuzu didn't want to make it work. The non pro controller I tried turning the vibration down with "Gear Button and D-pad Down" and it went from too strong vibration to kinda too weak.
> * Both Controllers connected in Windows X-Input mode now vibration works correctly for both controllers. I find I like it just a little bit stronger rumble so I press "Gear Button and D-pad Up" so it's on vibration level 3 for my preferred setting. :)
> 
> I only briefly tested Yuzu and Ruyjinx today with Crash Team Racing Nitro-Fueled. It seems again the way to get both controllers to work with vibration feedback is connected in Windows X-Input mode and only on Ryujinx emulator. Seriously I like the layout of Yuzu but I am really annoyed how much time I've wasted with bugs on Yuzu only to find and Ryujinx works better for most things.

I've never had issues with vibration on my King Kong 2's.

**Comment by Light-Blue-Yoshi** [comment]
Thanks I didn't consier that about the old outdated firmware. I'll try update the controlller CPU Firmware 1.17 now. I assume that if I'm on a really old version (let's say for example it was vers 1.09 or something then I only need to load the latest 1.17 to get all the previous updates included and up to 1.17? Like I don't have to go and find vers 1.16 and load that do I?

**Comment by OldManKain** [comment]
> Thanks I didn't consier that about the old outdated firmware. I'll try update the controlller CPU Firmware 1.17 now. I assume that if I'm on a really old version (let's say for example it was vers 1.09 or something then I only need to load the latest 1.17 to get all the previous updates included and up to 1.17? Like I don't have to go and find vers 1.16 and load that do I?

Yes all the new features will be in the latest updates, make sure you update both firmwares, 1.17 for CPU, and 4.9 for Firmware.

**Comment by Light-Blue-Yoshi** [comment]
Okay so I've updated the CPU Firmwares for both controllers now. When I went to run the update it said both controllers were 1.12 and it updated them to 1.17. Here's what happened when I tried to see if motion control was working or not in Ryujinx with Zelda - BotW. First time running Ruyjinx there was no motion control. I had do the same trick as before. I closed Ryujinx. Opened Yuzu. Made sure the controller was set in Yuzu, then oped Ryujinx and made sure the controller was set again in Ryujinx and then motion control worked. I tried closing Yuzu and Ryujinx and then restarting Ryujinx and motion control was still working. I turned off both controllers and restarted my Laptop and went straight into Ryujinx (without opening Yuzu first). Strange thing that motion control seems to still work! I don't know what to make of it. I did reset the gyro with the button combo listed above (also in the controller manual) on both controllers so I'll have to see how it goes in the next few days if I loose motion control and have to open Yuzu again to get Ryujinx to work with motion control or maybe it's fixed for good now?!?

P.S. During this time that motion control is working for Zelda I also tried out Mario Kart 8 Deluxe and found that both controllers again connected in Switch Pro Controller mode work with motion control (at least at the moment).

I tested with each controller in Windows X-Input mode and in Switch Pro Controller Mode. The difference in these two modes for  vibration is very significant in Mario Kart 8 Deluxe. I tested in Lightning Cup, Piranha Plant Pipeway racetrack. With the controllers on Windows X-Input they are much feel much nicer IMHO. There is a big range of difference in vibration intensity, for example if you turn around at the start and drive on the grass the vibration is only very mild but and mild to medium when power sliding on the dirt, then it's strong if you hit a wall or get smashed. On Switch Pro Controller mode everthing is the same really strong intensity so driving on the grass is really annoying as it rumbles just as much as hitting a wall or turbo boosting. Again played with the vibration ajustments on the controller. I turned the setting up one notch to level 3 when in Windows X-Input mode and left the controller in Switch Pro mode on the default level 2 vibration. They are then both about the same intensity when you hit a wall or get smashed by a green or red shell. I tried turning the vibration down when in Switch Pro Mode but it still makes everything from driving on grass, power sliding on dirt or hitting a wall or getting smashed still all the same intensity only just lowered to medium. I reccomend connecting one controller in Windows X-Input mode and one in Switch Pro mode and feeling the difference on the same track I tried.

**Comment by OldManKain** [comment]
> Okay so I've updated the CPU Firmwares for both controllers now. When I went to run the update it said both controllers were 1.12 and it updated them to 1.17. Here's what happened when I tried to see if motion control was working or not in Ryujinx with Zelda - BotW. First time running Ruyjinx there was no motion control. I had do the same trick as before. I closed Ryujinx. Opened Yuzu. Made sure the controller was set in Yuzu, then oped Ryujinx and made sure the controller was set again in Ryujinx and then motion control worked. I tried closing Yuzu and Ryujinx and then restarting Ryujinx and motion control was still working. I turned off both controllers and restarted my Laptop and went straight into Ryujinx (without opening Yuzu first). Strange thing that motion control seems to still work! I don't know what to make of it. I did reset the gyro with the button combo listed above (also in the controller manual) on both controllers so I'll have to see how it goes in the next few days if I loose motion control and have to open Yuzu again to get Ryujinx to work with motion control or maybe it's fixed for good now?!?
> 
> P.S. During this time that motion control is working for Zelda I also tried out Mario Kart 8 Deluxe and found that both controllers again connected in Switch Pro Controller mode work with motion control (at least at the moment).
> 
> I tested with each controller in Windows X-Input mode and in Switch Pro Controller Mode. The difference in these two modes for vibration is very significant in Mario Kart 8 Deluxe. I tested in Lightning Cup, Piranha Plant Pipeway racetrack. With the controllers on Windows X-Input they are much feel much nicer IMHO. There is a big range of difference in vibration intensity, for example if you turn around at the start and drive on the grass the vibration is only very mild but and mild to medium when power sliding on the dirt, then it's strong if you hit a wall or get smashed. On Switch Pro Controller mode everthing is the same really strong intensity so driving on the grass is really annoying as it rumbles just as much as hitting a wall or turbo boosting. Again played with the vibration ajustments on the controller. I turned the setting up one notch to level 3 when in Windows X-Input mode and left the controller in Switch Pro mode on the default level 2 vibration. They are then both about the same intensity when you hit a wall or get smashed by a green or red shell. I tried turning the vibration down when in Switch Pro Mode but it still makes everything from driving on grass, power sliding on dirt or hitting a wall or getting smashed still all the same intensity only just lowered to medium. I reccomend connecting one controller in Windows X-Input mode and one in Switch Pro mode and feeling the difference on the same track I tried.

Glad it's working for you now, but I do recommend updating the controller firmware too since there are two for the controller.  One for the CPU (Inside the controller), one for the Controller Itself.
https://www.gulikit.com/newsinfo/588841.html

**Comment by Light-Blue-Yoshi** [comment]
Yep I had already updated the Controller Firmware to 4.9 on both of them a little while ago so now that I've just updated the CPU Firmware on both of them today they should all be up to date now.

**Comment by OldManKain** [comment]
K I'll keep this open for a few days, if the problem doesn't come back just close it out with a comment that it's fixed.

**Comment by Light-Blue-Yoshi** [comment]
Thanks I'll keep testing it over the next few days to see if the problem comes back like before and I'll update here so anyone running into the same problem as I did can hopefully have the best solution.

**Comment by Light-Blue-Yoshi** [comment]
If anyone else has a GuliKit KingKong 2 or KingKong 2 Pro controller or an Official Nintendo Switch Pro Controller I'd love to here your feedback on if you've had the same issue listed in my original post.

**Comment by Squall-Leonhart** [comment]
This appears to be an issue on windows due to changes within SDL between 2.0.22 and 2.24.x, an issue on sdl's github should be getting reopened soon since it was now confirmed to affect more than Cemu.

**Comment by LoneBoco** [comment]
> If anyone else has a GuliKit KingKong 2 or KingKong 2 Pro controller or a genuine Nintendo Switch Pro Controller I'd love to hear your feedback on if you've had the same issue listed in my original post.

I have an official Pro controller and I've had my gyro stop working for periods of time too.

**Comment by MikePrime13** [comment]
I can confirm this -- I have to increase the gyro deadzone significantly, and this is present in both Pro Controller and Gulikit KingKong 2 Pro. I encountered the issue in both TOTK and SS when scoping in. To make sure this was not a hardware issue, I tried launching BOTW on CEMU and I had zero stick drift issues on CEMU.


> > > If anyone else has a GuliKit KingKong 2 or KingKong 2 Pro controller or a genuine Nintendo Switch Pro Controller I'd love to hear your feedback on if you've had the same issue listed in my original post.
> > 
> > 
> > I have an official Pro controller and I've had my gyro stop working for periods of time too.
> 
> Official Pro Controller here as well, gyro drifting right infinitely. This was also reported on #4760

**Comment by Light-Blue-Yoshi** [comment]
> I can confirm this -- I have to increase the gyro deadzone significantly, and this is present in both Pro Controller and Gulikit KingKong 2 Pro. I encountered the issue in both TOTK and SS when scoping in. To make sure this was not a hardware issue, I tried launching BOTW on CEMU and I had zero stick drift issues on CEMU.
> 
> > > > If anyone else has a GuliKit KingKong 2 or KingKong 2 Pro controller or a genuine Nintendo Switch Pro Controller I'd love to hear your feedback on if you've had the same issue listed in my original post.
> > > 
> > > 
> > > I have an official Pro controller and I've had my gyro stop working for periods of time too.
> > 
> > 
> > Official Pro Controller here as well, gyro drifting right infinitely. This was also reported on #4760

I'm reporting back to let everyone know that motion control aiming assist appears to still be working for me in Ryujinx version 1.1.777 and 1.1.799 with my GuliKit KingKong 2 Pro controller connected via bluetooth in Switch Pro Controller mode. This was tested after freshly booting my Windows 11 PC and loading Zelda BotW and also TotK. I have not opened Yuzu to kickstart the motion control in Ryujinx like I had to previously to get it working so I can only assume that me updating the GuliKit CPU Controller firmware from 1.12 to ver 1.17 may have helped solve this issue.

Thank you to the other people who have also replied back with their issues using their controllers. I also would like to mention that still have significant stick drift on Zelda BotW and TotK. When I load up a save game it doesn't happen every time but usually like 50% of he time, Link will be walking off to the left or right with the controller sticks not being touched. This is even now after updating all both the firmware modules on the controllers and also resetting the calibration on both the gyro and the sticks and triggers (using the button combos from the controller manual). What I've found seems to work every time for me in these two games is to simply roll the left in one full 360 degree circle. Link will run around in a circle and then magically the stick drift seems to go away and he no longer moves when the sticks aren't touched.

**Comment by LoneBoco** [comment]
Having to spin the sticks 360 degrees when you first start the game is an SDL thing I believe.  The sticks aren't calibrated initially so you have to spin them around to fully calibrate them.  As long as Ryujinx uses SDL for input I'm not entirely sure if it can be resolved.

**Comment by boymanguy** [comment]
Hi, I have a gulikit kingkongpro 2 and am having the same issue with gyro/motion drift. When using ryujinx, the gyro drifts slowly to the left. I tried adding a deadzone which mitigates the issue but the controller still turns more left than right while aiming in totk. If I move the controller up it still drifts left. It feels like I’m fighting with the controls when trying to aim. I have recalibrated the gyro on the controller many times but that is not the issue. This doesn’t happen for me in yuzu, cemu, or dolphin- only in ryujinx.

**Comment by RahuX** [comment]
Can also confirm this issue on my Gulikit KingKong Pro 2 controller. Controller constantly drifts to the right when gyro aiming as well as tries to reset to center.

The controller is on the latest firmwares available for it.

This also happens in Yuzu (EA or mainline), but **does not** happen in CEMU (latest 2.0 version).

Something else to note for comparison. I also have an 8bitdo Pro 2. It's on a very old firmware (1.05), however it does not experience gyro drift at all in any emulator.

**Comment by Squall-Leonhart** [comment]
replace the sdl2.dll in the ryujinx folder with the one from SDL 2.28.0RC (appears as 2.27.1 on the sdl releases page)

**Comment by Rarurin** [comment]
> replace the sdl2.dll in the ryujinx folder with the one from SDL 2.28.0RC (appears as 2.27.1 on the sdl releases page)

Doesn't change anything for me, same right drifting on gyro.

**Comment by Squall-Leonhart** [comment]
Drifting is an issue someone else attributed to this issue without any indication  of it actually being, only the ops issue and those with the ops issue have any relevance in this issue to this issue and of my reponse to the issue.

**Comment by OldManKain** [comment]
GuliKit has a new Firmware to "1. Strengthen the work stability of the gyroscope."
https://www.gulikit.com/newsinfo/588841.html
Also remember to set your GuliKit controller on a flat surface and calibrate it.

I personally have never had an issue with Gyro on my KingKong 2 non-pro (daily driver) but also tested my KingKong 2 pro and Zen Pro.

**Comment by felipecrs** [comment]
Make sure Steam is closed as well: https://www.reddit.com/r/yuzu/comments/yztix5/comment/jqjklnz/?utm_source=share&utm_medium=web2x&context=3

**Comment by ar-dead** [comment]
I can confirm that I have motion drift with my kinking 2 pro x2 and the switch pro controller however I do not have this issue with my 8bitdo wireless.  it's especially noticeable when activating ultrahand in totk.  it just drifts down and at a good pace too. setting a deadline to a large value defeats the purpose of motion controls as you now lack the precision.

I do not have any issues with any controller playing with Cemu or Yuzu... even replacing the sdl2.dll does not fix the issue.  which is sad cause Ruyjinx is the far better performance emulator for me as well as the best graphically thanks to SMAA high or SMAA Ultra, as well it being very user friendly I don't have to tweak and bob with each game like I do with Yuzu.  as well as the fact that I have no issues with cutscenes or upside down imagery with Ruyjinx where Yuzu this is an issue especially so with Streets of Rage 4.

**Comment by jxnding** [comment]
Also seeing this on Ryujinx with Zelda TOTK on Windows 10. I don't get any drift, just straight up lack of motion control. I also have a Dualshock 5 that works flawlessly on the same setup.

I have updated my GuliKit KK Pro 2 firmware per what @OldManKain said, but doesn't have an impact.
